File: hooks/post-tool-use/typescript_check_any_usage.py
#!/usr/bin/env python3
# /// script
# requires-python = ">=3.11"
# ///

from __future__ import annotations

import json
import re
import sys
from pathlib import Path
from typing import Any, TypedDict

RULES_DIR = Path(__file__).parent.parent.parent / "rules"


def _load_typescript_rules() -> dict[str, Any]:
    rules_file = RULES_DIR / "typescript.json"
    if rules_file.exists():
        try:
            with open(rules_file, encoding="utf-8") as f:
                return json.load(f)
        except (json.JSONDecodeError, OSError):
            pass
    return {}


_TS_RULES = _load_typescript_rules()
_TYPE_CONFIG = _TS_RULES.get("type_checking", {})
_MESSAGES = _TS_RULES.get("messages", {})


class EditOperation(TypedDict):
    old_string: str
    new_string: str


class WriteToolInput(TypedDict):
    file_path: str
    content: str


class EditToolInput(TypedDict):
    file_path: str
    old_string: str
    new_string: str


class MultiEditToolInput(TypedDict):
    file_path: str
    edits: list[EditOperation]


class PostToolUseInput(TypedDict):
    session_id: str
    tool_name: str
    transcript_path: str
    cwd: str
    hook_event_name: str
    tool_input: WriteToolInput | EditToolInput | MultiEditToolInput
    tool_response: dict[str, Any]


def main() -> None:
    """Main entry point for the PostToolUse hook."""
    hook_filename = Path(__file__).stem.replace("_", "-")
    print(f"\n[{hook_filename}]", file=sys.stderr)

    input_data = sys.stdin.read()
    if not input_data:
        sys.exit(0)

    try:
        data: PostToolUseInput = json.loads(input_data)
        tool_input = data["tool_input"]
    except (json.JSONDecodeError, KeyError):
        sys.exit(0)

    file_path = ""
    if "file_path" in tool_input:
        file_path = tool_input["file_path"]  # type: ignore[literal-required]

    if not file_path or not _is_valid_typescript_file(file_path):
        print(f"[typescript-any-check] Skipping: Not a TypeScript file: {file_path}")
        sys.exit(0)

    path = Path(file_path)
    if not path.exists():
        print(f"[typescript-any-check] Skipping: File does not exist: {file_path}")
        sys.exit(0)

    content = path.read_text()
    issues = check_any_usage(content, file_path)

    if not issues:
        print("[typescript-any-check] Success: No 'any' type usage detected")
        sys.exit(0)

    # Print error message and exit with code 2 to block the tool
    error_message = build_error_message(issues, file_path)
    print(error_message, file=sys.stderr)
    sys.exit(2)


def check_any_usage(content: str, file_path: str) -> list[dict[str, Any]]:
    """Check for 'any' type usage in TypeScript code.
    
    Returns list of issues found.
    """
    issues: list[dict[str, Any]] = []
    lines = content.split("\n")
    
    # Patterns to detect 'any' usage
    # Matches: `: any`, `<any>`, `Array<any>`, etc.
    # But NOT: 'company', 'many', comments, strings
    any_pattern = re.compile(
        r'''
        (?:^|[^a-zA-Z0-9_])  # Not preceded by alphanumeric or underscore
        (
            :\s*any\b|           # : any
            <any>|               # <any>
            <any,|               # <any,
            ,\s*any>|            # , any>
            Array<any>|          # Array<any>
            Promise<any>|        # Promise<any>
            \(\s*\)\s*:\s*any|   # () : any (function return)
            as\s+any\b           # as any
        )
        ''',
        re.VERBOSE | re.MULTILINE
    )
    
    for line_num, line in enumerate(lines, start=1):
        # Skip comments and strings (simple heuristic)
        stripped = line.strip()
        if stripped.startswith('//') or stripped.startswith('/*') or stripped.startswith('*'):
            continue
        
        # Check for @ts-ignore or @ts-expect-error on previous or same line
        has_ts_ignore = False
        if line_num > 1:
            prev_line = lines[line_num - 2].strip()
            has_ts_ignore = '@ts-ignore' in prev_line or '@ts-expect-error' in prev_line
        has_ts_ignore = has_ts_ignore or '@ts-ignore' in line or '@ts-expect-error' in line
        
        matches = any_pattern.finditer(line)
        for match in matches:
            # Skip if inside string literal (simple check)
            before_match = line[:match.start()]
            single_quotes = before_match.count("'") - before_match.count("\\'")
            double_quotes = before_match.count('"') - before_match.count('\\"')
            backticks = before_match.count('`') - before_match.count('\\`')
            
            if single_quotes % 2 != 0 or double_quotes % 2 != 0 or backticks % 2 != 0:
                continue  # Inside a string
            
            issues.append({
                'line': line_num,
                'column': match.start() + 1,
                'matched_text': match.group(1).strip(),
                'full_line': line.strip(),
                'has_ts_ignore': has_ts_ignore,
            })
    
    return issues


def build_error_message(issues: list[dict[str, Any]], file_path: str) -> str:
    """Build formatted error message for 'any' type usage."""
    message_parts = [
        "\n<typescript-any-usage-error>",
        f"**CRITICAL: 'any' type usage detected in {file_path}!**",
        "",
        "You are STRICTLY PROHIBITED from using the 'any' type in TypeScript code.",
        "The user has explicitly forbidden 'any' type usage.",
        "",
        f"Found {len(issues)} instance(s) of 'any' type usage:",
        "",
    ]
    
    for issue in issues:
        line = issue['line']
        matched = issue['matched_text']
        full_line = issue['full_line']
        has_ignore = issue['has_ts_ignore']
        
        ignore_note = " (has @ts-ignore/expect-error)" if has_ignore else ""
        message_parts.append(f"  Line {line}{ignore_note}: {matched}")
        message_parts.append(f"    {full_line}")
        message_parts.append("")
    
    message_parts.extend([
        "**You MUST fix this by:**",
        "1. Replace 'any' with 'unknown' if the type is truly unknown, then use type guards",
        "2. Use proper generic types (e.g., <T>, <T extends SomeType>)",
        "3. Use union types (e.g., string | number | null)",
        "4. Use specific interface or type definitions",
        "5. Use utility types (Partial<T>, Record<K, V>, etc.)",
        "",
        "**Examples of correct alternatives:**",
        "  ❌ const data: any = await fetch(...)",
        "  ✅ const data: unknown = await fetch(...)",
        "     if (typeof data === 'object' && data !== null) { ... }",
        "",
        "  ❌ function process(input: any): any { ... }",
        "  ✅ function process<T>(input: T): T { ... }",
        "  ✅ function process(input: unknown): string | number { ... }",
        "",
        "  ❌ const items: any[] = []",
        "  ✅ const items: Item[] = []",
        "  ✅ const items: Array<string | number> = []",
        "",
        "</typescript-any-usage-error>\n",
    ])
    
    return "\n".join(message_parts)


def _is_valid_typescript_file(file_path: str) -> bool:
    if not file_path:
        return False

    valid_extensions = set(_TS_RULES.get("extensions", [".ts", ".tsx", ".mts", ".cts"]))
    excluded_extensions = _TS_RULES.get("excluded_extensions", [".d.ts"])
    path = Path(file_path)

    if path.suffix not in valid_extensions:
        return False

    for excluded in excluded_extensions:
        if file_path.endswith(excluded):
            return False

    return True


if __name__ == "__main__":
    main()

File: hooks/post-tool-use/typescript-check-any-usage.ts
#!/usr/bin/env bun
import { readFileSync, existsSync } from "fs";
import { join, dirname } from "path";

const HOOK_NAME = "typescript-check-any-usage";
const RULES_DIR = join(dirname(dirname(__dirname)), "rules");

interface LanguageRules {
  extensions?: string[];
  excluded_extensions?: string[];
  messages?: Record<string, string | string[]>;
}

function loadTypeScriptRules(): LanguageRules {
  const rulesFile = join(RULES_DIR, "typescript.json");
  if (existsSync(rulesFile)) {
    try {
      return JSON.parse(readFileSync(rulesFile, "utf-8"));
    } catch {
      return {};
    }
  }
  return {};
}

const rules = loadTypeScriptRules();

const DEFAULT_EXTENSIONS = [".ts", ".tsx", ".mts", ".cts"];
const DEFAULT_EXCLUDED = [".d.ts"];

const ANY_KEYWORD = "any";
const ANY_PATTERNS = [
  { pattern: new RegExp(":\\s*" + ANY_KEYWORD + "\\b", "g"), desc: "Direct '" + ANY_KEYWORD + "' type annotation" },
  { pattern: new RegExp("<" + ANY_KEYWORD + ">", "g"), desc: "Generic '" + ANY_KEYWORD + "' type parameter" },
  { pattern: new RegExp("as\\s+" + ANY_KEYWORD + "\\b", "g"), desc: "Type assertion to '" + ANY_KEYWORD + "'" },
  { pattern: new RegExp("Array<" + ANY_KEYWORD + ">", "g"), desc: "Array of '" + ANY_KEYWORD + "'" },
  { pattern: new RegExp("Promise<" + ANY_KEYWORD + ">", "g"), desc: "Promise of '" + ANY_KEYWORD + "'" },
];

interface AnyUsage {
  line: number;
  column: number;
  text: string;
  description: string;
}

interface PostToolUseInput {
  tool_input: {
    file_path?: string;
    notebook_path?: string;
  };
}

async function main(): Promise<void> {
  console.error(`\n[${HOOK_NAME}]`);

  const input = await Bun.stdin.text();
  const filePath = getTargetFilePath(input);

  if (!filePath) {
    console.log(`[${HOOK_NAME}] Skipping: No valid TypeScript file`);
    process.exit(0);
  }

  const content = readFileSync(filePath, "utf-8");
  const violations = detectAnyUsage(content);

  if (violations.length === 0) {
    console.log(`[${HOOK_NAME}] Success: No '${ANY_KEYWORD}' type usage found`);
    process.exit(0);
  }

  const message = buildErrorMessage(violations, filePath);
  console.error(message);
  process.exit(2);
}

function getTargetFilePath(inputData: string): string | null {
  if (!inputData) return null;

  try {
    const data = JSON.parse(inputData) as PostToolUseInput;
    const toolInput = data.tool_input;

    const filePath = toolInput.file_path || toolInput.notebook_path || "";
    return isValidTypeScriptFile(filePath) ? filePath : null;
  } catch {
    return null;
  }
}

function isValidTypeScriptFile(filePath: string): boolean {
  if (!filePath) return false;

  const validExtensions = rules.extensions || DEFAULT_EXTENSIONS;
  const excludedExtensions = rules.excluded_extensions || DEFAULT_EXCLUDED;

  const ext = filePath.substring(filePath.lastIndexOf("."));

  if (!validExtensions.includes(ext)) return false;

  for (const excluded of excludedExtensions) {
    if (filePath.endsWith(excluded)) return false;
  }

  return existsSync(filePath);
}

function detectAnyUsage(content: string): AnyUsage[] {
  const violations: AnyUsage[] = [];
  const lines = content.split("\n");

  for (let i = 0; i < lines.length; i++) {
    const line = lines[i];
    if (!line) continue;

    if (line.includes("// @ts-ignore") || line.includes("// @ts-expect-error")) {
      continue;
    }

    for (const { pattern, desc } of ANY_PATTERNS) {
      pattern.lastIndex = 0;
      let match;

      while ((match = pattern.exec(line)) !== null) {
        violations.push({
          line: i + 1,
          column: match.index + 1,
          text: line.trim().substring(0, 80),
          description: desc,
        });
      }
    }
  }

  return violations;
}

function buildErrorMessage(violations: AnyUsage[], filePath: string): string {
  const alternativesRaw = rules.messages?.any_alternatives;
  const alternatives = Array.isArray(alternativesRaw) ? alternativesRaw : [
    "Use 'unknown' if the type is truly unknown, then use type guards",
    "Use proper generic types (e.g., <T>, <T extends SomeType>)",
    "Use union types (e.g., string | number | null)",
    "Use specific interface or type definitions",
  ];

  let msg = `
<typescript-any-usage-error>
CRITICAL: '${ANY_KEYWORD}' TYPE USAGE DETECTED

File: ${filePath}
Found ${violations.length} violation(s)

`;

  for (const v of violations) {
    msg += `Line ${v.line}: ${v.description}\n  ${v.text}\n\n`;
  }

  msg += `
You are STRICTLY PROHIBITED from using the '${ANY_KEYWORD}' type.

ALTERNATIVES:
`;

  for (const alt of alternatives) {
    msg += `- ${alt}\n`;
  }

  msg += `
FIX IMMEDIATELY: Replace all '${ANY_KEYWORD}' types with proper type annotations.
</typescript-any-usage-error>
`;

  return msg;
}

main().catch((err) => {
  console.error(`[${HOOK_NAME}] Error:`, err);
  process.exit(0);
});
